x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/import_rules_utils.ts

```diff
@@ -46,116 +50,111 @@ export interface RuleExceptionsPromiseFromStreams {
-                    statusCode: 409,
-                    message: `rule_id: "${parsedRule.rule_id}" already exists`,
-                  })
-                );
-              }
-            } catch (err) {
-              resolve(
-                createBulkErrorObject({
-                  ruleId: parsedRule.rule_id,
-                  statusCode: err.statusCode ?? 400,
-                  message: err.message,
-                })
-              );
-            }
-          } catch (error) {
-            reject(error);
-          }
+  const importRuleResponses = [...rulesResponseAcc];
+  const ruleChunks = chunk(IMPORT_RULES_CHUNK_SIZE, rules);
+  const importRule = async (ruleToImport: PromiseFromStreams): Promise<ImportRuleResponse> => {
+    if (ruleToImport instanceof Error) {
+      // If the JSON object had a validation or parse error then we return
+      // early with the error and an (unknown) for the ruleId
+      return createBulkErrorObject({
+        statusCode: 400,
+        message: ruleToImport.message,
+      });
+    }
+
+    try {
+      const [exceptionErrors, exceptions] = checkRuleExceptionReferences({
+        rule: ruleToImport,
+        existingLists,
+      });
+
+      for (const exceptionError of exceptionErrors) {
+        importRuleResponses.push(exceptionError);
+      }
+
+      throwAuthzError(await mlAuthz.validateRuleType(ruleToImport.type));
+      const existingRule = await readRules({
+        rulesClient,
+        ruleId: ruleToImport.rule_id,
+        id: undefined,
+      });
+
+      if (existingRule != null && !overwriteRules) {
+        return createBulkErrorObject({
+          ruleId: ruleToImport.rule_id,
+          statusCode: 409,
+          message: `rule_id: "${ruleToImport.rule_id}" already exists`,
         });
-        return [...accum, importsWorkerPromise];
-      }, [])
-    );
-    importRuleResponse = [...importRuleResponse, ...newImportRuleResponse];
+      }
+
+      if (existingRule == null) {
+        await createRules({
+          rulesClient,
+          params: {
+            ...ruleToImport,
+            exceptions_list: [...exceptions],
+          },
+          allowMissingConnectorSecrets,
+        });
+
+        return {
+          rule_id: ruleToImport.rule_id,
+          status_code: 200,
+        };
+      }
+
+      await updateRules({
+        rulesClient,
+        existingRule,
+        ruleUpdate: {
+          ...ruleToImport,
+          exceptions_list: [...exceptions],
+        },
+        allowMissingConnectorSecrets,
+      });
+
+      return {
+        rule_id: ruleToImport.rule_id,
+        status_code: 200,
+      };
+    } catch (err) {
+      return createBulkErrorObject({
+        ruleId: ruleToImport.rule_id,
+        statusCode: err.statusCode ?? 400,
+        message: err.message,
+      });
+    }
+  };
+
+  for (const rulesChunk of ruleChunks) {
+    const chunkImportRuleResponses = await Promise.all(rulesChunk.map(wrapInMacrotask(importRule)));
```

Comment:
Just an idea: should we run each chunk in a macrotask instead of each rule?

Alternative idea: should we use promise pool instead of chunking approach here?

Reply:
I ended up wrapping rules chunk import in a macro task. Promise pool doesn't allow to do that without modification of its implementation.